From 037f8d0febec6a5a4f2426e0fed5db0eed7a31b1 Mon Sep 17 00:00:00 2001
From: Bill Hlavacek <bill.hlavacek@gmail.com>
Date: Thu, 15 Oct 2026 13:18:27 -0600
Subject: [PATCH] bngsim: throw instead of exit(1) in the DOR reaction classes
 and observables

DORRxnClass, DOR2RxnClass, Observable and SpeciesObservable called
exit(1) on an error, so a model NFsim cannot run killed the host process
with no message. BNG-written models reach two of them at load:

- A(b)%x + B(a)%y -> A(b!1).B(a!1) fA(x)*fB(y), with fA and fB reading
  Molecules observables: the local-function rate law reads both
  reactants, which DORRxnClass does not support.
- The same rule with Species observables: a local function cannot read a
  Species observable, because SpeciesObservable::clone() was never written
  ("not yet updated ... fix me").

bngsim's Python process exited with code 1 during initialize() on both.
Throw std::runtime_error naming the rule or observable and the reason, as
carry 0005 did for six other files. Load-time throws are caught by
NFinput::initReactionRules and NFinput::initFunctions, which fail the load,
so the host reports them; the internal-error paths at run time
(Observable::subtract, rate-factor changes on a non-DOR reactant) now
raise rather than exit. The ReactionClass destructor that runs when a
derived constructor throws frees only arrays the base constructor has
already allocated. (bngsim issue #854 review)
---
 src/NFcore/observable.cpp                 |  9 ++++---
 src/NFreactions/reactions/DORreaction.cpp | 33 ++++++++++++-----------
 2 files changed, 22 insertions(+), 20 deletions(-)

diff --git a/src/NFcore/observable.cpp b/src/NFcore/observable.cpp
index fe6f1d0..7725596 100644
--- a/src/NFcore/observable.cpp
+++ b/src/NFcore/observable.cpp
@@ -1,5 +1,6 @@
 #include <iostream>
 #include "observable.hh"
+#include <stdexcept>
 
 
 using namespace std;
@@ -79,7 +80,7 @@ void Observable::subtract()
 	if(count==0){
 		cerr << "Error in observable count!! Removing from an empty observable!"
 		     << "Observable named: " << obsName << endl;
-		exit(1);
+		throw std::runtime_error("Observable " + obsName + ": removing from an empty observable");
 	}
 
 	count--;
@@ -103,7 +104,7 @@ void Observable::subtract( int n_matches )
 	{
 		cerr << "Error in observable count!! Removing " << n_matches << " matches will result in a negative match count!"
 		     << "Observable named: " << obsName << endl;
-		exit(1);
+		throw std::runtime_error("Observable " + obsName + ": removing matches would make the count negative");
 	}
 
 	// First, we subtract from our observable count
@@ -463,7 +464,7 @@ SpeciesObservable::~SpeciesObservable()
 Observable * SpeciesObservable::clone() {
 	vector <TemplateMolecule *> tmList;
 	cout<<"in clone species observable, this is not yet updated to handle stoch observables.  fix me."<<endl;
-	exit(1);
+	throw std::runtime_error("SpeciesObservable " + obsName + ": a local function cannot read a Species observable (NFsim cannot clone one)");
 	for(int t=0; t<n_templates; t++)
 		tmList.push_back(templateMolecules[t]);
 	return 0; //new SpeciesObservable(obsName+"_clone",tmList);
@@ -473,7 +474,7 @@ int SpeciesObservable::isObservable(Molecule *m) const
 {
 	cerr<<"Comparing a Species observable '"<<obsName<<"' to a molecule!"<<endl;
 	cerr<<"You can only compare Species observable to a complexes!  Quitting."<<endl;
-	exit(1);
+	throw std::runtime_error("SpeciesObservable " + obsName + ": compared to a molecule; a Species observable matches complexes only");
 
 }
 
diff --git a/src/NFreactions/reactions/DORreaction.cpp b/src/NFreactions/reactions/DORreaction.cpp
index 238e9ff..c115ab5 100644
--- a/src/NFreactions/reactions/DORreaction.cpp
+++ b/src/NFreactions/reactions/DORreaction.cpp
@@ -2,6 +2,7 @@
 
 
 #include "reaction.hh"
+#include <stdexcept>
 
 
 
@@ -38,7 +39,7 @@ DORRxnClass::DORRxnClass(
 					{   // DOR reactant is a population!
 						cout<<"Error when creating DORRxnClass: "<<name<<endl;
 						cout<<"DOR reactant cannot be a population type."<<endl;
-						exit(1);
+						throw std::runtime_error("DORRxnClass " + name + ": the DOR reactant cannot be a population type");
 					}
 
 					DORreactantIndex=r;
@@ -47,7 +48,7 @@ DORRxnClass::DORRxnClass(
 					cout<<"Error when creating DORRxnClass: "<<name<<endl;
 					cout<<"DOR reactions currently only support one DOR reactant.  This means that you can"<<endl;
 					cout<<"only have a pointer to one or the other of the two reactants, but not both."<<endl;
-					exit(1);
+					throw std::runtime_error("DORRxnClass " + name + ": a local-function rate law can read only one of the reactants, not both");
 				}
 			}
 		}
@@ -56,7 +57,7 @@ DORRxnClass::DORRxnClass(
 		cout<<"Error when creating DORRxnClass: "<<name<<endl;
 		cout<<"You don't have any pointers onto the Molecules or Species, so you can't have a local function!"<<endl;
 		cout<<"That means that this is not a DOR reaction at all!"<<endl;
-		exit(1);
+		throw std::runtime_error("DORRxnClass " + name + ": the local-function rate law reads no reactant, so this is not a DOR reaction");
 	}
 
 
@@ -72,7 +73,7 @@ DORRxnClass::DORRxnClass(
 	if((unsigned)function->getNumOfArgs()!=lfArgumentPointerNameList.size()) {
 		cout<<"Error when creating DORRxnClass: "<<name<<endl;
 		cout<<"Number of arguments and LocalFunctionArgumentPointerNameList size do not match!"<<endl;
-		exit(1);
+		throw std::runtime_error("DORRxnClass " + name + ": the function's argument count does not match its argument pointers");
 	}
 
 
@@ -109,7 +110,7 @@ DORRxnClass::DORRxnClass(
 			cout<<"Error when creating DOR reaction: "<<name<<endl;
 			cout<<"Could not find a match in the templateMolecules for the pointer reference to species/molecule: ";
 			cout<<lfArgumentPointerNameList.at(i)<<endl;
-			exit(1);
+			throw std::runtime_error("DORRxnClass " + name + ": no template molecule matches a local-function argument");
 		}
 	}
 
@@ -470,7 +471,7 @@ double DORRxnClass::pickLocalFunctionParameter(MappingSet* ms, int index, Molecu
 		}
 		// there is truly no possible mapping. User mistake prob, the error message could use somem improvement
 		cout<<"Internal error in LocalFunction::evaluateOn()! Trying to evaluate a function with unknown scope."<<endl;
-		exit(1);
+		throw std::runtime_error("DORRxnClass " + name + ": internal error evaluating a local function with an unknown argument");
 
 }
 //This function takes a given mappingset and looks up the value of its local
@@ -736,7 +737,7 @@ void DORRxnClass::notifyRateFactorChange(Molecule * m, int reactantIndex, int rx
 	} else {
 		cout<<"Internal Error in DORRxnClass::notifyRateFactorChange!!  : trying to change a rate\n";
 		cout<<"factor of a non-DOR reactant.  That means this function was called in error!\n";
-		exit(1);
+		throw std::runtime_error("DORRxnClass " + name + ": internal error, rate factor change on a non-DOR reactant");
 	}
 }
 
@@ -806,7 +807,7 @@ DOR2RxnClass::DOR2RxnClass(
 					{   // DOR reactant is a population!
 						cout<<"Error when creating DOR2RxnClass: "<<name<<endl;
 						cout<<"DOR reactant1 cannot be a population type."<<endl;
-						exit(1);
+						throw std::runtime_error("DOR2RxnClass " + name + ": DOR reactant 1 cannot be a population type");
 					}
 
 					DORreactantIndex1 = r;
@@ -819,7 +820,7 @@ DOR2RxnClass::DOR2RxnClass(
 						// DOR reactant is a population!
 						cout<<"Error when creating DOR2RxnClass: "<<name<<endl;
 						cout<<"DOR reactant2 cannot be a population type."<<endl;
-						exit(1);
+						throw std::runtime_error("DOR2RxnClass " + name + ": DOR reactant 2 cannot be a population type");
 					}
 
 					DORreactantIndex2 = r;
@@ -830,7 +831,7 @@ DOR2RxnClass::DOR2RxnClass(
 				else {
 					cout<<"Error when creating DOR2RxnClass: "<<name<<endl;
 					cout<<"DOR2 reactions only support two DOR reactants."<<endl;
-					exit(1);
+					throw std::runtime_error("DOR2RxnClass " + name + ": DOR2 reactions support only two DOR reactants");
 				}
 			}
 		}
@@ -839,7 +840,7 @@ DOR2RxnClass::DOR2RxnClass(
 	if (DORreactantIndex1==-1 || DORreactantIndex2==-1) {
 		cout<<"Error when creating DOR2RxnClass: "<<name<<endl;
 		cout<<"DOR2RxnClass requires pointers into two different reactant patterns, but fewer than 2 were found!"<<endl;
-		exit(1);
+		throw std::runtime_error("DOR2RxnClass " + name + ": the rate law must point into two different reactant patterns");
 	}
 
 
@@ -858,7 +859,7 @@ DOR2RxnClass::DOR2RxnClass(
 	if((unsigned)function1->getNumOfArgs()!=lfArgumentPointerNameList1.size()) {
 		cout<<"Error when creating DOR2RxnClass: "<<name<<endl;
 		cout<<"Number of arguments in function1 and LocalFunctionArgumentPointerList1 size do not match!"<<endl;
-		exit(1);
+		throw std::runtime_error("DOR2RxnClass " + name + ": function 1's argument count does not match its argument pointers");
 	}
 
 	n_argMolecules1=lfArgumentPointerNameList1.size();
@@ -891,7 +892,7 @@ DOR2RxnClass::DOR2RxnClass(
 			cout<<"Error when creating DOR2 reaction: "<<name<<endl;
 			cout<<"Could not find a match in the templateMolecules for a pointer reference to species/molecule: ";
 			cout<<lfArgumentPointerNameList1.at(i)<<endl;
-			exit(1);
+			throw std::runtime_error("DOR2RxnClass " + name + ": no template molecule matches a function 1 argument");
 		}
 	}
 
@@ -919,7 +920,7 @@ DOR2RxnClass::DOR2RxnClass(
 	if((unsigned)function2->getNumOfArgs()!=lfArgumentPointerNameList2.size()) {
 		cout<<"Error when creating DOR2RxnClass: "<<name<<endl;
 		cout<<"Number of arguments in function2 and LocalFunctionArgumentPointerList2 size do not match!"<<endl;
-		exit(1);
+		throw std::runtime_error("DOR2RxnClass " + name + ": function 2's argument count does not match its argument pointers");
 	}
 
 	n_argMolecules2=lfArgumentPointerNameList2.size();
@@ -952,7 +953,7 @@ DOR2RxnClass::DOR2RxnClass(
 			cout<<"Error when creating DOR2 reaction: "<<name<<endl;
 			cout<<"Could not find a match in the templateMolecules for a pointer reference to species/molecule: ";
 			cout<<lfArgumentPointerNameList2.at(i)<<endl;
-			exit(1);
+			throw std::runtime_error("DOR2RxnClass " + name + ": no template molecule matches a function 2 argument");
 		}
 	}
 
@@ -1529,7 +1530,7 @@ void DOR2RxnClass::notifyRateFactorChange(Molecule * m, int reactantIndex, int r
 	else {
 		cout<<"Internal Error in DORRxnClass::notifyRateFactorChange!!  : trying to change a rate\n";
 		cout<<"factor of a non-DOR reactant.  That means this function was called in error!\n";
-		exit(1);
+		throw std::runtime_error("DOR2RxnClass " + name + ": internal error, rate factor change on a non-DOR reactant");
 	}
 }
 
-- 
2.54.0 (Apple Git-157)

